API/models/rbac.go

```diff
@@ -0,0 +1,122 @@
+package models
+
+import (
+	u "p3/utils"
+	"strings"
+
+	"go.mongodb.org/mongo-driver/bson"
+	"go.mongodb.org/mongo-driver/bson/primitive"
+)
+
+// Roles
+const (
+	Manager = "manager"
+	User    = "user"
+	Viewer  = "viewer"
+)
+
+// Actions
+const (
+	READ = iota
+	WRITE
+	READONLYNAME
+)
+
+const ROOT_DOMAIN = "*"
+
+func CheckDomainExists(domain string) bool {
+	if domain == ROOT_DOMAIN {
+		return true
+	}
+	x, e := GetEntity(bson.M{"hierarchyName": domain}, "domain", u.RequestFilters{}, nil)
+	return e == "" && x != nil
+}
+
+func GetRequestFilterByDomain(userRoles map[string]string) (bson.M, bool) {
+	filter := bson.M{}
+	if userRoles[ROOT_DOMAIN] == Manager || userRoles[ROOT_DOMAIN] == User {
+		return filter, true
+	}
+	domainPattern := ""
+	for domain, role := range userRoles {
+		switch role {
+		case User:
+		case Manager:
+			if domainPattern == "" {
+				domainPattern = domain
+			} else {
+				domainPattern = domainPattern + "|" + domain
+			}
+		}
+	}
+	if domainPattern == "" {
+		return filter, false
+	} else {
+		filter["domain"] = primitive.Regex{Pattern: domainPattern, Options: ""}
+		return filter, true
+	}
+}
+
+func CheckUserPermissions(userRoles map[string]string, objEntity, requestType int, objDomain string) (bool, int) {
+	if objEntity == u.DOMAIN {
+		if userRoles[ROOT_DOMAIN] == Manager {
+			return true, WRITE
+		}
+		for userDomain, role := range userRoles {
+			if role == Manager && strings.Contains(objDomain, userDomain) {
```

Comment:
userDomain should not only be contained in objDomain, it should be a prefix,
otherwise someone who has access to the domain b, will have access to the domain a.b